optuna/integration/dask.py

```diff
@@ -0,0 +1,733 @@
+import asyncio
+import datetime
+import inspect
+from typing import Any
+from typing import Callable
+from typing import Container
+from typing import Dict
+from typing import Generator
+from typing import Iterable
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Tuple
+from typing import Union
+import uuid
+
+import optuna
+from optuna._experimental import experimental_class
+from optuna._imports import try_import
+from optuna.distributions import BaseDistribution
+from optuna.distributions import distribution_to_json
+from optuna.distributions import json_to_distribution
+from optuna.storages import BaseStorage
+from optuna.study import StudyDirection
+from optuna.study._frozen import FrozenStudy
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+with try_import() as _imports:
+    import distributed
+    from distributed.utils import thread_state
+    from distributed.worker import get_client
+
+# Serialization utilities
+
+
+def _serialize_datetime(dt: datetime.datetime) -> dict:
+    return {"__datetime__": True, "as_str": dt.strftime("%Y%m%dT%H:%M:%S.%f")}
+
+
+def _deserialize_datetime(data: dict) -> datetime.datetime:
+    return datetime.datetime.strptime(data["as_str"], "%Y%m%dT%H:%M:%S.%f")
```

Comment:
[nit] It might be good to use `dt.isoformat(timespec="microseconds")` and `datetime.fromisoformat()` here if there is no any restrictions in Dask, since `optuna.storages.JournalStorage` also uses it to serialize datetime objects.

(Please note that `datetime.fromisoformat` is not available in Python 3.6, so the following code is required until #4150 is merged.)
https://github.com/optuna/optuna/blob/596acc9812db9272fa85fc5db5f532ff9ba33e76/optuna/storages/_journal/storage.py#L48-L50

Reply:
Sure, I don't see an issue with using `isoformat` / `fromisoformat`

Reply:
Switched to using `datetime.fromisoformat` since https://github.com/optuna/optuna/pull/4150 has been merged